odiglet/pkg/ebpf/manager.go

```diff
@@ -0,0 +1,371 @@
+	instance "github.com/odigos-io/odigos/k8sutils/pkg/instrumentation_instance"
+	workloadUtils "github.com/odigos-io/odigos/k8sutils/pkg/workload"
+	"github.com/odigos-io/odigos/odiglet/pkg/detector"
+	"github.com/odigos-io/odigos/odiglet/pkg/kube/utils"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrContainerNotInPodSpec    = errors.New("container not found in pod spec")
+	ErrContainerNameNotReported = errors.New("container name not reported in environment variables")
+	ErrDeviceNotDetected        = errors.New("device not detected")
+	ErrNoInstrumentationFactory = errors.New("no ebpf factory found")
+)
+
+type InstrumentationStatusReason string
+
+const (
+	FailedToLoad       InstrumentationStatusReason = "FailedToLoad"
+	FailedToInitialize InstrumentationStatusReason = "FailedToInitialize"
+	LoadedSuccessfully InstrumentationStatusReason = "LoadedSuccessfully"
+	FailedToRun        InstrumentationStatusReason = "FailedToRun"
+)
+
+type InstrumentationHealth bool
+
+const (
+	InstrumentationHealthy   InstrumentationHealth = true
+	InstrumentationUnhealthy InstrumentationHealth = false
+)
+
+type configUpdate struct {
+	workloadKey types.NamespacedName
+	config      *odigosv1.InstrumentationConfig
+}
+
+type ConfigUpdateFunc func(ctx context.Context, workloadKey types.NamespacedName, config *odigosv1.InstrumentationConfig) error
+
+type Manager struct {
+	ProcEvents    <- chan detector.ProcessEvent
+	Client        client.Client
+	Factories     map[FactoryID]Factory
+	Logger        logr.Logger
+	DetailsByPid  map[int]InstrumentationDetails
+	configUpdates chan configUpdate
+
+	done chan struct{}
+	stop context.CancelFunc
+}
+
+func NewManager(client client.Client, logger logr.Logger, factories map[FactoryID]Factory, procEvents chan detector.ProcessEvent) *Manager {
+	return &Manager{
+		ProcEvents:    procEvents,
+		Client:        client,
+		Factories:     factories,
+		Logger:        logger,
+		DetailsByPid:  make(map[int]InstrumentationDetails),
+		configUpdates: make(chan configUpdate),
+		done:          make(chan struct{}),
+	}
+}
+
+func (m *Manager) UpdateConfig(ctx context.Context, workloadKey types.NamespacedName, config *odigosv1.InstrumentationConfig) error {
+	// send a config update event for the specified workload
+	select {
+	case m.configUpdates <- configUpdate{workloadKey: workloadKey, config: config}:
+		return nil
+	case <-ctx.Done():
+		if ctx.Err() == context.DeadlineExceeded {
+			return errors.New("failed to update config of workload: timeout waiting for config update")
+		}
+		return ctx.Err()
+	}
+}
+
+func (m *Manager) Run(ctx context.Context) {
+	defer close(m.done)
+
+	runLoopCtx, stop := context.WithCancel(ctx)
+	m.stop = stop
+
+	// main event loop for handling instrumentations
+	for {
```

Comment:
I am not sure how or if we should handle it, but I think this design has a noisy neighbor issue:
Imaging a Pod with the following bash script:
```bash
while true; do
    echo "test"
done
```
This will trigger tons of ProcEvents and will cause starvation for all other pods / configuration changes.
What do you think? cc @blumamir

Reply:
aka Odiglet DDos